franklin/github/views.py

```diff
@@ -55,6 +63,152 @@ def callback(request):
+#                url = response.links['next']['url']
+#                have_next_page = True
+#
+#    if not repos:
+#        logger.error('Failed to find repos for owner', owner_name)
+#    return repos
+
+def get_site_from_request(request):
+    owner_name = request.data.get('owner', None)
+    owner_id = request.data.get('owner_id', None)
+    repo_name = request.data.get('repository', None)
+    repo_id = request.data.get('repository_id', None)
+    if owner_id and repo_id:
+        site, created = Site.objects.get_or_create(owner_id=owner_id, 
+                                                   repo_name_id=repo_id)
+        if site:
+            site.owner = owner_name
+            site.repo_name = repo_name
+            site.oauth_token = os.environ['GITHUB_OAUTH']
+            site.save()
+            return site
+    return None
+
+
+def get_franklin_config(site):
+    url = github_base + 'repos/' + site.owner + '/' + site.repo_name + '/contents/.franklin.yml'
+    #TODO - This will fetch the file from the default master branch
+    headers = {
+                'content-type': 'application/json',
+                'Authorization': 'token ' + site.oauth_token
+              }
+    response = make_rest_get_call(url, headers)
+
+    if response is not None:
+        download_url = response.json().get('download_url', None)
+        raw_data = make_rest_get_call(download_url, None)
+        # TODO - validation and cleanup needed here similar to:
+        # http://stackoverflow.com/a/22231372
+        franklin_config = yaml.load(raw_data.text)
+        return franklin_config
+    return None
+
+def create_repo_webhook(site):
+    # TODO - check for existing webhook and update if needed (or skip)
+    
+    # TODO - Confirm that a header token is the best/most secure way to go
+    headers = {
+                'content-type': 'application/json',
+                'Authorization': 'token ' + site.oauth_token
+              }
+    body = {
+                'name': 'web',
+                'events': ['push'],
+                'active': True,
+                'config': {
+                                'url': base_url + 'deployed/',
+                                'content_type': 'json'
+                          }
+            }
+    url = github_base + 'repos/' + site.owner + '/' + site.repo_name + '/hooks'
+    response = make_rest_post_call(url, headers, body)
+    if response:
+        return True
+    return False
+
+@api_view(['POST'])
+def register_repo(request):
+    # TODO - Lock this endpoint down so it's only callable from the future
+    # admin panel.
+    if request.method == 'POST':
+        site = get_site_from_request(request)
+        if site:
+            config = get_franklin_config(site)
+            if config:
+                # update DB with any relevant .franklin config itmes here.
+
+                # TODO - We should return the error we get from github if we
+                # fail. Refactor this.
```

Comment:
Implementing this will require refactoring the GET/POST methods above. I think I'll do this after this code is merged. Definitely want to return more robust errors than just `400_BAD_REQUEST` for everything that goes wrong. Especially for these kind of endpoints that rely heavily on external dependencies to be available and responding correctly. (in this case, github)